pkg/store/bucket.go

```diff
@@ -1574,6 +1574,12 @@ func (s *BucketStore) LabelNames(ctx context.Context, req *storepb.LabelNamesReq
 	s.mtx.RUnlock()
 
 	if err := g.Wait(); err != nil {
+		if statusErr, ok := status.FromError(err); ok {
+			if statusErr.Code() == codes.ResourceExhausted {
+				return nil, status.Error(codes.ResourceExhausted, err.Error())
+			}
+		}
+
 		return nil, status.Error(codes.Internal, err.Error())
```

Comment:
Maybe we can just relay the actual code like we do in `Series()` call?
```suggestion
		code := codes.Internal
		if s, ok := status.FromError(errors.Cause(err)); ok {
			code = s.Code()
		}
		return nil, status.Error(code, err.Error())
```